errs/error.go

```diff
@@ -0,0 +1,249 @@
+package errs
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"github.com/pkg/errors"
+)
+
+// StatusCoder interface is used by errors that returns the HTTP response code.
+type StatusCoder interface {
+	StatusCode() int
+}
+
+// StackTracer must be by those errors that return an stack trace.
+type StackTracer interface {
+	StackTrace() errors.StackTrace
+}
+
+// Option modifies the Error type.
+type Option func(e *Error) error
+
+// WithMessage returns an Option that modifies the error by overwriting the
+// message only if it is empty.
+func WithMessage(format string, args ...interface{}) Option {
+	return func(e *Error) error {
+		if len(e.Msg) > 0 {
+			return e
+		}
+		e.Msg = fmt.Sprintf(format, args...)
+		return e
+	}
+}
+
+// Error represents the CA API errors.
+type Error struct {
+	Status int
+	Err    error
+	Msg    string
+}
+
+// New returns a new Error. If the given error implements the StatusCoder
+// interface we will ignore the given status.
+func New(status int, err error, opts ...Option) error {
+	var e *Error
+	if sc, ok := err.(StatusCoder); ok {
+		e = &Error{Status: sc.StatusCode(), Err: err}
+	} else {
+		cause := errors.Cause(err)
+		if sc, ok := cause.(StatusCoder); ok {
+			e = &Error{Status: sc.StatusCode(), Err: err}
+		} else {
+			e = &Error{Status: status, Err: err}
+		}
+	}
+	for _, o := range opts {
+		o(e)
+	}
+	return e
+}
+
+// ErrorResponse represents an error in JSON format.
+type ErrorResponse struct {
+	Status  int    `json:"status"`
+	Message string `json:"message"`
+}
+
+// Cause implements the errors.Causer interface and returns the original error.
+func (e *Error) Cause() error {
+	return e.Err
+}
+
+// Error implements the error interface and returns the error string.
+func (e *Error) Error() string {
+	return e.Err.Error()
+}
+
+// StatusCode implements the StatusCoder interface and returns the HTTP response
+// code.
+func (e *Error) StatusCode() int {
+	return e.Status
+}
+
+// Message returns a user friendly error, if one is set.
+func (e *Error) Message() string {
+	if len(e.Msg) > 0 {
+		return e.Msg
+	}
+	return e.Err.Error()
+}
+
+// Wrap returns an error annotating err with a stack trace at the point Wrap is
+// called, and the supplied message. If err is nil, Wrap returns nil.
+func Wrap(status int, e error, m string, opts ...Option) error {
+	if e == nil {
+		return nil
+	}
+	if err, ok := e.(*Error); ok {
+		err.Err = errors.Wrap(err.Err, m)
+		e = err
+	} else {
+		e = errors.Wrap(e, m)
+	}
+	return StatusCodeError(status, e, opts...)
```

Comment:
Passing the options to `StatusCodeError` that does not support all the status codes looks weird, especially if we want to add a new one. I think you should just create the Error here.